texar/modules/pretrained/xlnet_model_utils.py

```diff
@@ -0,0 +1,393 @@
+
+    @staticmethod
+    def default_hparams():
+        return {
+            "name": "relative_multihead_attention",
+            "initializer": None,
+            "num_heads": 16,
+            "hidden_dim": 1024,
+            "head_dim": 64,
+            "dropout": 0.1,
+            "attention_dropout": 0.1,
+            "use_segments": True,
+        }
+
+    @staticmethod
+    def _rel_shift(x, klen=-1):
+        """Perform relative shift to form the relative attention score."""
+        x_size = tf.shape(x)
+
+        x = tf.reshape(x, [x_size[1], x_size[0], x_size[2], x_size[3]])
+        x = tf.slice(x, [1, 0, 0, 0], [-1, -1, -1, -1])
+        x = tf.reshape(x, [x_size[0], x_size[1] - 1, x_size[2], x_size[3]])
+        x = tf.slice(x, [0, 0, 0, 0], [-1, klen, -1, -1])
+
+        return x
+
+    def _compute_attention_score(self, q_head, k_head_h, v_head_h, k_head_r,
+                                 segment_mat, attn_mask=None, **kwargs):
+        mode = kwargs.get("mode", global_mode())
+
+        # Content based attention score.
+        q_head_rw = q_head + self.r_w_bias
+        # attn_ac: (seq_len, tot_len, batch_size, n_head)
+        attn_ac = tf.einsum('ibnd,jbnd->ijbn', q_head_rw, k_head_h)
+
+        # Position based attention score.
+        q_head_rr = q_head + self.r_r_bias
+        # attn_bd: (seq_len, tot_len, batch_size, n_head)
+        attn_bd = tf.einsum('ibnd,jbnd->ijbn', q_head_rr, k_head_r)
+        attn_bd = self._rel_shift(attn_bd, klen=tf.shape(attn_ac)[1])
+
+        # Segment based attention score.
+        if segment_mat is None:
+            attn_ef = 0
+        else:
+            q_head_rs = q_head + self.r_s_bias
+            attn_ef = tf.einsum(
+                'ibnd,snd->ibns', q_head_rs, self.segment_embed)
+            attn_ef = tf.einsum('ijbs,ibns->ijbn', segment_mat, attn_ef)
+
+        # Merge attention scores and perform masking.
+        # attn_score: (seq_len, tot_len, batch_size, n_head)
+        attn_score = (attn_ac + attn_bd + attn_ef) * self.scale
+        if attn_mask is not None:
+            # attn_score = attn_score * (1 - attn_mask) - 1e30 * attn_mask
+            attn_score = attn_score - 1e30 * attn_mask
+
+        # attention probability
+        attn_prob = tf.nn.softmax(attn_score, 1)
+        attn_prob = self.dropout(attn_prob,
+                                 training=(mode == tf.estimator.ModeKeys.TRAIN))
+
+        # attention output
+        attn_vec = tf.einsum('ijbn,jbnd->ibnd', attn_prob, v_head_h)
+
+        return attn_vec
+
+    def _post_attention(self, attn_vec):
+        shape = tf.shape(attn_vec)
+        attn_vec = tf.reshape(
+            tensor=attn_vec,
+            shape=[shape[0], shape[1], self.num_heads*self.head_dim])
+        attn_out = self.output_projection(attn_vec)
+        attn_out = self.dropout(attn_out)
+        return attn_out
+
+    def _build(self, states_h, pos_embed, states_g=None, segment_mat=None,
```

Comment:
Add docstring to all `_build` methods, on par w/ `forward` in pytorch

Reply:
There should be an explicit `mode` argument instead of `**kwargs` which is harder to read.

Reply:
Done! [9280366]

Reply:
docstring to all `_build` to be added

Reply:
Changes are visible in f578727